### PR TITLE
feat(comp: notification): add notification component

### PR DESCRIPTION
**Pull request by lo56ve**
re #545

## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows [our guidelines](https://github.com/IDuxFE/idux/blob/main/packages/site/src/docs/Contributing.zh.md#commit)
- [ ] Tests for the changes have been added/updated or not needed
- [x] Docs and demo have been added/updated or not needed

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Component style update
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Application (the showcase website) / infrastructure changes
- [ ] Other... Please describe:

### CODE REVIEW

File: packages/components/notification/src/NotificationProvider.tsx [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+  notifications: NotificationOptions[],
+  initPlacementOffset: NotificationPlacementOffset,
+) {
+  const realPlacement = notifications?.[0]?.placement ?? props.placement ?? config.placement
+  const realOffset = initPlacementOffset[realPlacement]
+  return getPosition(realOffset, realPlacement)
+}
+
+function getPosition(offset: NotificationProps['offset'], placement: NotificationPlacement) {
+  const { verticalOffset, horizontalOffset } = getDirectionOffset(offset)
```
Comment (danranVm) [inline packages/components/notification/src/NotificationProvider.tsx:261]:
```suggestion
 const offsets = isArray(offset) ? offset : [offset, offset]
  const [ verticalOffset, horizontalOffset ] = offsets.map(convertCssPixel)
```

File: packages/components/notification/src/NotificationProvider.tsx [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+  return notifications.map(item => {
+    const { key, visible = true } = item
+    const onClose = () => destroy(key!)
+    return <Notification {...item} onClose={onClose} visible={visible}></Notification>
+  })
+}
+
+function getMoveName(placement: NotificationPlacement) {
+  const moveNameMap = {
+    topEnd: 'ix-move-end',
```
Comment (danranVm) [inline packages/components/notification/src/NotificationProvider.tsx:302]: ix 前缀通过 globalConfig 获取

File: packages/components/notification/src/NotificationProvider.tsx [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+    getPlacementInitValue<NotificationProps['offset']>(() => 0),
+  )
+  watch(
+    placementNotifications,
+    (newNotifications, oldNotifications) => {
+      if (newNotifications && oldNotifications) {
+        Object.entries(newNotifications).forEach(([item, notifications]) => {
+          const placement = item as NotificationPlacement
+          if (notifications.length === 1 && oldNotifications[placement].length === 0) {
+            // 同个方位的已第一个的offset为基准
```
Comment (danranVm) [inline packages/components/notification/src/NotificationProvider.tsx:238]: 这个设定感觉不太合适。。。

File: packages/components/notification/src/NotificationProvider.tsx [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+  config: NotificationConfig,
+  notifications: Ref<NotificationOptions[]>,
+) {
+  return computed(() => {
+    const initValue = getPlacementInitValue<NotificationOptions[]>(() => []) as NotificationPlacementMap
+    return notifications.value.reduce((acc, item) => {
+      const notificationProps = {
+        ...getRealObj(omit<NotificationConfig, 'maxCount'>(config, ['maxCount'])),
+        ...getRealObj(omit<NotificationProviderProps, 'maxCount'>(props, ['maxCount'])),
+        ...getRealObj(omit<NotificationOptions, 'onDestroy'>(item, ['onDestroy'])),
```
Comment (danranVm) [inline packages/components/notification/src/NotificationProvider.tsx:110]: 这种写法会造成一些不必要的响应式依赖，另外可读性也。。

File: packages/components/notification/src/NotificationProvider.tsx [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+    return ref
+  }
+
+  const { info, success, warning, error } = notificationType.reduce(
+    (acc, type) => ({
+      ...acc,
+      [type]: (options: Omit<NotificationOptions, 'type'>) => open({ ...options, type }),
+    }),
+    {} as Record<NotificationType, (options: Omit<NotificationOptions, 'type'>) => NotificationRef>,
+  )
```
Comment (danranVm) [inline packages/components/notification/src/NotificationProvider.tsx:186]:
```suggestion
  const  notificationTypes = ['info', 'success', 'warning', 'error'] as const
  const [info, success, warning, error] = notificationTypes.map(type => {
    return (options?: Omit<NotificationOptions, 'type'>) => open({ ...options, type })
  })
```
要沉迷 reduce 呢，map 不香吗？

File: packages/components/notification/src/Notification.tsx [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+  onBeforeUnmount(() => clearTimer())
+
+  return { visible, close, onMouseEnter, onMouseLeave }
+}
+
+function getIconNode(icon: NotificationProps['icon']) {
+  return isString(icon) ? <IxIcon name={icon}></IxIcon> : icon
+}
+
+function getNode<T extends NotificationProps, K extends NotificationNodePropKey>(
```
Comment (danranVm) [inline packages/components/notification/src/Notification.tsx:144]: 这个函数有点过度封装了，即影响了可读性，也降低了执行效率（每次都会去创建 footer Node）。

File: packages/components/notification/src/NotificationProvider.tsx [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+  return getPosition(realOffset, realPlacement)
+}
+
+function getPosition(offset: NotificationProps['offset'], placement: NotificationPlacement) {
+  const { verticalOffset, horizontalOffset } = getDirectionOffset(offset)
+  const notificationPosMap = new Map<NotificationPlacement, NotificationPosition>([])
+    .set('topStart', { top: verticalOffset, left: horizontalOffset })
+    .set('topEnd', { top: verticalOffset, right: horizontalOffset })
+    .set('bottomStart', { bottom: verticalOffset, left: horizontalOffset })
+    .set('bottomEnd', { bottom: verticalOffset, right: horizontalOffset })
```
Comment (danranVm) [inline packages/components/notification/src/NotificationProvider.tsx:266]: const  positions = {
topStart: ['top', 'left'],
topEnd: ['top', 'right'],
bottomStart: ['bottom', 'left'],
bottomEnd: ['bottom', 'right'],
} as const // 定义到外面

const [ verticalPosition, horizontalPosition ] = positions[placement ]
return { [verticalPosition]:verticalOffset, [horizontalPosition]:horizontalOffset   }

**Review by danranVm — CHANGES_REQUESTED** [review]

File: packages/components/notification/src/Notification.tsx [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+import { useGlobalConfig } from '@idux/components/config'
+import { IxIcon } from '@idux/components/icon'
+import { IxSpace } from '@idux/components/space'
+
+import { notificationProps } from './types'
+
+const clsPrefix = 'ix-notification'
+
+const defaultIconTypes = {
+  success: 'check-circle',
```
[the remark this answers is not in the document]
Comment (lo56ve) [reply packages/components/notification/src/Notification.tsx:27]: 声明 // staticIcons.ts  的作用是？

[the remark this answers is not in the document]
Comment (lo56ve) [reply packages/components/notification/src/Notification.tsx:27]: > 

声明 // staticIcons.ts 的作用是？

[the remark this answers is not in the document]
Comment (lo56ve) [reply packages/components/notification/src/Notification.tsx:27]: > 这个直接作为 defaultConfig 把 另外注意在 // staticIcons.ts 里声明一下。

声明 // staticIcons.ts 的作用是？

[the remark this answers is not in the document]
Comment (lo56ve) [reply packages/components/notification/src/Notification.tsx:27]: fix

File: packages/components/notification/src/Notification.tsx [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+  return { visible, close, onMouseEnter, onMouseLeave }
+}
+
+function getIconNode(icon: NotificationProps['icon']) {
+  return isString(icon) ? <IxIcon name={icon}></IxIcon> : icon
+}
+
+function getNode<T extends NotificationProps, K extends NotificationNodePropKey>(
+  props: T,
+  slots: SetupContext['slots'],
```
[the remark this answers is not in the document]
Comment (lo56ve) [reply packages/components/notification/src/Notification.tsx]: fix

File: packages/components/notification/docs/Index.zh.md [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+
+全局的通知提醒配置
+
+| 名称             | 说明                           | 类型                                                      | 默认值     | 全局配置 | 备注                                                         |
+| ---------------- | ------------------------------ | --------------------------------------------------------- | ---------- | -------- | ------------------------------------------------------------ |
+| `maxCount`       | 同一时间可展示的最大提示数量   | `number`                                                  | 5          | ☑️        | -                                                            |
+| `destroyOnHover` | 鼠标悬浮时是否允许销毁         | `boolean`                                                 | `false`    | ☑️        | -                                                            |
+| `duration`       | 自动销毁的延时，单位毫秒       | `number`                                                  | 4500       | ☑️        | -                                                            |
+| `icon`           | 自定义通知图标                 | `string \| VNode`                                         | -          | -        | -                                                            |
```
[the remark this answers is not in the document]
Comment (lo56ve) [reply packages/components/notification/docs/Index.zh.md]: fix

File: packages/components/style/themes/default.less [hunk 11]
```diff
@@ -414,6 +414,38 @@
 @message-wrapper-top: 15%;
 @message-wrapper-zindex: @zindex-l4-1;
 
+/* notification -------------------------- */
```
[the remark this answers is not in the document]
Comment (lo56ve) [reply packages/components/style/themes/default.less]: fix

File: packages/components/notification/src/NotificationProvider.tsx [hunk 12]
```diff
@@ -0,0 +1,322 @@
...
+
+    provide(notificationProviderToken, apis)
+    expose(apis)
+
+    return () => {
+      return (
+        <>
+          {slots.default?.()}
+          <CdkPortal target="ix-notification-container" load={loadContainer.value}>
+            {placementGroup.value}
```
[the remark this answers is not in the document]
Comment (lo56ve) [reply packages/components/notification/src/NotificationProvider.tsx]: fix

File: packages/components/notification/src/NotificationProvider.tsx [hunk 13]
```diff
@@ -0,0 +1,322 @@
...
+          </CdkPortal>
+        </>
+      )
+    }
+  },
+})
+
+// 将通知按位置分类
```
[the remark this answers is not in the document]
Comment (lo56ve) [reply packages/components/notification/src/NotificationProvider.tsx]: fix

File: packages/components/notification/src/NotificationProvider.tsx (hunk 1 above)
Comment (lo56ve) [reply packages/components/notification/src/NotificationProvider.tsx]: fix

File: packages/components/notification/src/NotificationProvider.tsx (hunk 3 above)
Comment (lo56ve) [reply packages/components/notification/src/NotificationProvider.tsx]: fix

File: packages/components/notification/src/NotificationProvider.tsx (hunk 5 above)
Comment (lo56ve) [reply packages/components/notification/src/NotificationProvider.tsx]: fix

File: packages/components/notification/src/NotificationProvider.tsx (hunk 2 above)
Comment (lo56ve) [reply packages/components/notification/src/NotificationProvider.tsx]: fix

File: packages/components/notification/src/NotificationProvider.tsx (hunk 4 above)
Comment (lo56ve) [reply packages/components/notification/src/NotificationProvider.tsx]: fix

File: packages/components/notification/src/Notification.tsx (hunk 6 above)
Comment (lo56ve) [reply packages/components/notification/src/Notification.tsx:144]: fix

File: packages/components/notification/src/NotificationProvider.tsx (hunk 7 above)
Comment (lo56ve) [reply packages/components/notification/src/NotificationProvider.tsx]: fix

File: packages/components/notification/src/Notification.tsx [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+  },
+})
+
+function useClasses(props: NotificationProps) {
+  return computed(() => [clsPrefix, `${clsPrefix}-${props.type}`])
+}
+
+function useIcon(props: NotificationProps) {
+  return {
+    isShowIcon: computed(() => !!props.type),
```
[the remark this answers is not in the document]
Comment (lo56ve) [reply packages/components/notification/src/Notification.tsx]: fix

File: packages/components/notification/src/Notification.tsx [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+
+import { callEmit, useControlledProp } from '@idux/cdk/utils'
+import { IxButton } from '@idux/components/button'
+import { useGlobalConfig } from '@idux/components/config'
+import { IxIcon } from '@idux/components/icon'
+import { IxSpace } from '@idux/components/space'
+
+import { notificationProps } from './types'
+
+const clsPrefix = 'ix-notification'
```
[the remark this answers is not in the document]
Comment (lo56ve) [reply packages/components/notification/src/Notification.tsx]: fix

**Pull request merged**